lua/neotest/consumers/state/internal.lua

```diff
@@ -0,0 +1,198 @@
+local au_result = "NeoTestStatus"
+local au_running = "NeoTestStarted"
+
+---@class state
+---@field private _result table<integer, table<string, neotest.Tree> >
+---@field private _status table
+---@field private _cache table
+---@field private _running table<string, table<string, string>>
+---@field private _adapters table<integer, string>
+---@field private _client neotest.Client
+state = {
+  _result = {},
+  _status = {},
+  _cache = {},
+  _running = {},
+  _adapters = {},
+}
+
+---Escape characters that have a special meaning in patterns
+---@param text string
+---@return string
+local function escape(text)
+  text = text:gsub("%-", "%%-")
+  text = text:gsub("%.", "%%.")
+  return text
+end
+
+---Update the internal list of currently running tests
+---@param adapter_id string
+---@param position_ids string
+function state:_update_running(adapter_id, position_ids)
+  self._running[position_ids] = {
+    adapter = string.match(adapter_id, "(.-):"),
+    path = position_ids,
+  }
+
+  -- Trigger Event
+  vim.api.nvim_exec_autocmds(
+    "User",
+    { pattern = au_running, data = { id = adapter_id, running = position_ids } }
+  )
+end
+
+---Process results and count pass/fail rates, save to cache
+-- Passed, Failed, Skipped
+---@param results table<string, neotest.Result>
+function state:_process_result(results)
+  for path, data in pairs(results) do
+    local test_path = string.match(path, "^([^:]*)::")
+    if test_path ~= nil then
+      local test_name = string.match(path, escape(test_path) .. "::(.*)")
+      self:_set_status(test_path, test_name, data.status)
+    end
+  end
+  self:_update_cache()
+end
+
+---Update status for specific test
+---@param path string
+---@param status string
+function state:_set_status(path, name, status)
+  if self._status[path] == nil then
+    self._status[path] = {}
+  end
+  self._status[path][name] = status
```

Comment:
No names are not guaranteed to be unique. For example in the rspec adapter, it uses line numbers in the IDs because the names can be anything including empty. Also for adapters where names do have to be unique, it's only unique in their namespaces, e.g. in python two test classes in the same file can define a test with the same name